go/mysqlconn/server.go

```diff
@@ -168,72 +168,53 @@ func (l *Listener) handle(conn net.Conn, connectionID uint32) {
-		if !l.authServer.UseClearText() {
-			userData, err := l.authServer.ValidateHash(salt, user, authResponse)
-			if err != nil {
-				c.writeErrorPacketFromError(err)
-				return
-			}
-			c.User = user
-			c.UserData = userData
-			// We're good.
-			break
-		}
+	// See what auth method the AuthServer wants to use for that user.
+	authServerMethod, err := l.authServer.AuthMethod(user)
+	if err != nil {
+		c.writeErrorPacketFromError(err)
+		return
+	}
 
-		// Our AuthServer cannot use mysql_native_password, it
-		// needs the real password.  Let's request that.
-		renegotiateWithClearText = true
-	case mysqlClearPassword:
-		// Client sent us a clear text password. Let's use it if we can.
-		if !l.AllowClearTextWithoutTLS && c.Capabilities&CapabilityClientSSL == 0 {
-			c.writeErrorPacket(CRServerHandshakeErr, SSUnknownSQLState, "Cannot use clear text authentication over non-SSL connections.")
-			return
-		}
-		userData, err := l.authServer.ValidateClearText(user, string(authResponse))
+	// Compare with what the client sent back.
+	switch {
+	case authServerMethod == MysqlNativePassword && authMethod == MysqlNativePassword:
+		// Both server and client want to use MysqlNativePassword:
+		// the negotiation can be completed right away, using the
+		// ValidateHash() method.
+		userData, err := l.authServer.ValidateHash(salt, user, authResponse)
 		if err != nil {
 			c.writeErrorPacketFromError(err)
 			return
 		}
 		c.User = user
 		c.UserData = userData
-		break
+
+	case authServerMethod == MysqlNativePassword:
+		// The server really wants to use MysqlNativePassword,
+		// but the client returned a result for something else:
+		// not sure this can happen, so not supporting this now.
+		c.writeErrorPacket(CRServerHandshakeErr, SSUnknownSQLState, "Client asked for auth %v, but server wants auth mysql_native_password", authMethod)
+		return
+
 	default:
-		// Client decided to use something we don't understand.
-		// Let's try again with clear text password.
-		renegotiateWithClearText = true
-	}
+		// The server wants to use something else, re-negotiate.
 
-	// If we need to re-negotiate with clear text, do it.
-	if renegotiateWithClearText {
-		// Check error conditions.
+		// The negotiation happens in clear text. Let's check we can.
 		if !l.AllowClearTextWithoutTLS && c.Capabilities&CapabilityClientSSL == 0 {
 			c.writeErrorPacket(CRServerHandshakeErr, SSUnknownSQLState, "Cannot use clear text authentication over non-SSL connections.")
 			return
 		}
 
-		if err := c.writeAuthSwitchRequest(mysqlClearPassword, nil); err != nil {
+		// Switch our auth method to what the server wants.
+		// Note: for now don't support the extra data.
+		if err := c.writeAuthSwitchRequest(authServerMethod, nil); err != nil {
 			log.Errorf("Error write auth switch packet for client %v: %v", c.ConnectionID, err)
 			return
 		}
 
-		// The client is supposed to just send the data in a single packet.
-		// It is a zero-terminated string.
```

Comment:
I was mistaken, it's appropriately asking for the dialog.so. My test was bad because my client was caching the file handle of the so, since i moved it after starting the client. I've now verified that at least the dialog.so is being requested appropriately in the re-negotiation.